cirq/ops/raw_types_test.py

```diff
@@ -459,6 +459,100 @@ def test_tagged_operation():
+                                       exponent=1.0,
+                                       connected=True,
+                                       exponent_qubit_index=None,
+                                       auto_exponent_parens=True)
+    args = cirq.CircuitDiagramInfoArgs(None, None, None, None, None, False)
+    assert cirq.circuit_diagram_info(tagged_h) == expected
+    assert (cirq.circuit_diagram_info(tagged_h,
+                                      args) == cirq.circuit_diagram_info(h))
+
+    c = cirq.Circuit(tagged_h)
+    diagram_with_tags = "(1, 1): ───H['tag1']───"
+    diagram_without_tags = "(1, 1): ───H───"
+    assert str(cirq.Circuit(tagged_h)) == diagram_with_tags
+    assert c.to_text_diagram() == diagram_with_tags
+    assert c.to_text_diagram(include_tags=False) == diagram_without_tags
+
+
+def test_circuit_diagram_tagged_global_phase():
+    # Tests global phase operation
+    q = cirq.NamedQubit('a')
+    global_phase = cirq.GlobalPhaseOperation(coefficient=-1.0).with_tags('tag0')
+
+    assert (cirq.circuit_diagram_info(global_phase,
+                                      default='default') == 'default')
+    cirq.testing.assert_has_diagram(cirq.Circuit(global_phase),
+                                    "\n\nglobal phase:   π['tag0']",
+                                    use_unicode_characters=True)
+    cirq.testing.assert_has_diagram(cirq.Circuit(global_phase),
+                                    "\n\nglobal phase:   π",
+                                    use_unicode_characters=True,
+                                    include_tags=False)
+
+    expected = cirq.CircuitDiagramInfo(wire_symbols=(),
+                                       exponent=1.0,
+                                       connected=True,
+                                       exponent_qubit_index=None,
+                                       auto_exponent_parens=True)
+
+    # Operation with no qubits and returns diagram info with no wire symbols
+    class NoWireSymbols(cirq.GlobalPhaseOperation):
+
+        def _circuit_diagram_info_(self, args: 'cirq.CircuitDiagramInfoArgs'
+                                  ) -> 'cirq.CircuitDiagramInfo':
+            return expected
+
+    no_wire_symbol_op = NoWireSymbols(coefficient=-1.0).with_tags('tag0')
+    assert (cirq.circuit_diagram_info(no_wire_symbol_op,
+                                      default='default') == expected)
+    cirq.testing.assert_has_diagram(cirq.Circuit(no_wire_symbol_op),
+                                    "\n\nglobal phase:   π['tag0']",
+                                    use_unicode_characters=True)
+
+    tag1 = cirq.GlobalPhaseOperation(coefficient=1j).with_tags('tag1')
+    tag2 = cirq.GlobalPhaseOperation(coefficient=1j).with_tags('tag2')
+    c = cirq.Circuit([cirq.X(q), tag1, tag2])
+    cirq.testing.assert_has_diagram(c,
+                                    """\
+a: ─────────────X───────────────────
+
+global phase:   π['tag1', 'tag2']""",
```

Comment:
Do you think combining tags is the right way to go?  What if they are both tagged tag1?  Should this be `['tag1', 'tag1']`?

Reply:
That's how I have it implemented currently.  I don't know.  I think it's the most sensible option.  At least this gives you an indication that there are two tagged operations hiding in the moment and distinguishes it from one being tagged and the other not being tagged.  

Having two (redundant) GlobalPhaseOperations in the same moment both tagged with the same tags and displaying it in a circuit diagram is a goofy case, so I am not sure it's worth worrying about too much.

Reply:
I agree.  I forgot that many global phases could go into one moment.  So what does the output look like with a global phase in two different moments?